clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -6809,3 +6811,57 @@ ExprResult SemaSYCL::ActOnUniqueStableNameExpr(SourceLocation OpLoc,
 
   return BuildUniqueStableNameExpr(OpLoc, LParen, RParen, TSI);
 }
+
+void SemaSYCL::performSYCLDelayedAttributesAnalaysis(const FunctionDecl *FD) {
+
+  if (UserProvidedSYCLKernelFunctions.contains(FD))
+    return;
+
+  if (const SYCLReqdWorkGroupSizeAttr *Attr =
+          FD->getAttr<SYCLReqdWorkGroupSizeAttr>()) {
+    bool IsDependent = false;
+    for (const auto *CE : {dyn_cast<ConstantExpr>(Attr->getXDim()),
+                           dyn_cast_or_null<ConstantExpr>(Attr->getYDim()),
+                           dyn_cast_or_null<ConstantExpr>(Attr->getZDim())})
+      IsDependent = IsDependent ||
+                    (CE && (CE->isValueDependent() || CE->isTypeDependent()));
+    if (!IsDependent)
+      Diag(Attr->getLoc(),
+           diag::warn_sycl_incorrect_use_attribute_non_kernel_function)
+          << Attr;
+  }
+
+  if (const IntelReqdSubGroupSizeAttr *Attr =
+          FD->getAttr<IntelReqdSubGroupSizeAttr>()) {
+    bool IsDependent = false;
+    if (const auto *CE = dyn_cast<ConstantExpr>(Attr->getValue()))
+      IsDependent = (CE && (CE->isValueDependent() || CE->isTypeDependent()));
+
+    if (!IsDependent)
+      Diag(Attr->getLoc(),
+           diag::warn_sycl_incorrect_use_attribute_non_kernel_function)
+          << Attr;
+  }
+
+  if (const SYCLWorkGroupSizeHintAttr *Attr =
+          FD->getAttr<SYCLWorkGroupSizeHintAttr>()) {
+    bool IsDependent = false;
+    for (const auto *CE : {dyn_cast<ConstantExpr>(Attr->getXDim()),
+                           dyn_cast_or_null<ConstantExpr>(Attr->getYDim()),
+                           dyn_cast_or_null<ConstantExpr>(Attr->getZDim())})
+      IsDependent = IsDependent ||
+                    (CE && (CE->isValueDependent() || CE->isTypeDependent()));
+    if (!IsDependent)
+      Diag(Attr->getLoc(),
+           diag::warn_sycl_incorrect_use_attribute_non_kernel_function)
+          << Attr;
+  }
```

Comment:
The repetitive stuff above could be refactored.

Reply:
Okay, I will refactor it